chore(docs) Add proxy-tunnel docs and small api/typescript tidy up (#748)
electric-sql/electric: docs/api/cli.md

---
title: CLI & Generator
description: >-
  Command-line interface for the `npx electric-sql generate` script and other commands.
sidebar_position: 30
---

# CLI Commands

The Electric client library comes with a CLI providing a couple of helpful commands for building Electric apps:

> - [`generate`](#generate) - The client library generator; this command builds the client library for your applications to use.

Reviewer: We usually refer to using the generator to generate a type-safe client. So a more appropriate wording here would be `The type-safe client generator`, IMO. Saying that it builds "the client library" is makes its relationship with the actual Typescript client library more confusing.

> - [`proxy-tunnel`](#proxy-tunnel) - A tool to enable you to connect to the Electric Postgres Proxy when you can only open an http port for the sync later.

Reviewer: This phrasing looks somewhat clunky. I would describe it as a tool to enable you to connect to the Electric migrations proxy when it's deployed behind an restrictive firewall that only lets HTTP traffic through. Or, alternatively, when it's running on a hosting provider that only allows incoming HTTP connections.

| <span className="no-wrap">`--proxy`</span> | `<url>` | Provides the url to connect to Electric's database proxy. If not provided, it uses the url set in the `ELECTRIC_PROXY_URL` environment variable. If that variable is not set, it resorts to the default url which is <span class="break-all">`postgresql://prisma:proxy_password@localhost:65432/electric`</span>. |
| <span className="no-wrap">`--out`</span> | `<path>` | Specifies where to output the generated client. Defaults to `./src/generated/client` |
| <span className="no-wrap">`--watch`</span> | `<pollingInterval>` | Run the generator in watch mode. Accepts an optional polling interval (in milliseconds) which defines how often to poll Electric for database migrations. The default polling interval is 1000ms. |

Note that the `--watch` flag can be used with or without a polling interval:

```shell
npx electric-sql generate --watch
# or with an explicit polling interval
npx electric-sql generate --watch 5000
```

## `proxy-tunnel`

> Some deployment options for Electric only allow you to expose a single port, in which case only the http port, and not the [Postgres Proxy](../usage/data-modelling/migrations#migrations-proxy) port, will be available to connect to. In order to enable connecting to run migrations and use the generate command in these setups, you can enable a special "Proxy Tunnel" that tunnels the Postgres Proxy TCP connection over a Websocket to the Electric sync service. This is enabled on the sync service by setting the environment variable `PG_PROXY_PORT=http`.

Reviewer: The phrase "Some deployment options for Electric" feels like putting the blame on Electric. I would express it in a different manner:
> Some hosting providers only allow HTTP connections which poses a challenge for deploying Electric to their platforms since it uses a separate port for connections to the migrations proxy.

> The `npx electric-sql proxy-tunnel` command is provided to forward TCP traffic from your local machine to the Electric Postgres Proxy when it is in this mode. It binds to a local port, allowing you to use the generator command, perform migrations, and connect with psql.

Reviewer: ```suggestion
The `npx electric-sql proxy-tunnel` command is provided to forward TCP traffic from your local machine to the Electric Postgres Proxy when it has tunneling enabled. It binds to a local port, allowing you to use the generator command, perform migrations, and connect with psql.
```
